### PR TITLE
feat(nvmx): initial support of device events

### PR DESCRIPTION
**Pull request by mtzaurus**
BlockDevice now supports registration of callbacks to notify about
device events (as of now only device removal event is supported).
In addition, nvmf tests now use dynamically launched Mayastor
instances.

Implements CAS-670

### CODE REVIEW

**Review by Hoverbear — APPROVED** [review]

File: mayastor/src/bdev/dev/nvmx/controller.rs [hunk 1]
```diff
@@ -96,22 +106,35 @@ pub struct NvmeControllerInner<'a> {
     adminq_poller: poller::Poller<'a>,
     io_device: Arc<IoDevice>,
 }
+
+type EventCallbackList = Vec<fn(DeviceEventType, &str)>;
```
Comment (gila) [inline mayastor/src/bdev/dev/nvmx/controller.rs:110]: Should we limit the number of entries we can insert? (or maybe a different data structure) or are the number of events, not a concern? If they *might* be, I think we should add an issue for it to look into it later.

File: mayastor/src/bdev/dev/nvmx/controller.rs [hunk 2]
```diff
@@ -502,6 +529,38 @@ impl<'a> NvmeController<'a> {
...
+            (*l)(event, &self.name);
+        }
+        listeners.len()
+    }
+
+    /// Register listener to monitor device events related to this controller.
+    pub fn add_event_listener(
+        &self,
+        listener: fn(DeviceEventType, &str),
+    ) -> Result<(), CoreError> {
```
Comment (gila) [inline mayastor/src/bdev/dev/nvmx/controller.rs:554]: nit; we never generate an error here.

File: mayastor/src/bdev/dev/nvmx/device.rs [hunk 3]
```diff
@@ -119,6 +120,17 @@ impl NvmeBlockDevice {
             name: String::from(name),
         }
     }
+
+    fn get_controller(
+        &self,
+    ) -> Result<Arc<Mutex<NvmeController<'static>>>, CoreError> {
```
Comment (gila) [inline mayastor/src/bdev/dev/nvmx/device.rs:126]: 'static? 🤔

**Review by gila — COMMENTED** [review]
few minor questions, thanks @mtzaurus 👍

File: mayastor/src/bdev/dev/nvmx/device.rs (hunk 3 above)
Comment (Hoverbear) [reply mayastor/src/bdev/dev/nvmx/device.rs:126]: Do we want the returned NVMe controller to last possibly the entire life of the binary run?

File: mayastor/src/bdev/dev/nvmx/controller.rs (hunk 1 above)
Comment (mtzaurus) [reply mayastor/src/bdev/dev/nvmx/controller.rs:110]: Sure thing. I will create a ticket to consider improvement in device events storing.

File: mayastor/src/bdev/dev/nvmx/controller.rs (hunk 2 above)
Comment (mtzaurus) [reply mayastor/src/bdev/dev/nvmx/controller.rs:554]: I wanted to keep the API more flexible. Later on we might add checks against the maximum limit of event listeners, etc.

File: mayastor/src/bdev/dev/nvmx/device.rs (hunk 3 above)
Comment (mtzaurus) [reply mayastor/src/bdev/dev/nvmx/device.rs:126]: Indeed we can move controller lookup close to the caller and get rid of get_controller(), as its used only once.

**Comment by mtzaurus** [comment]
Merged manually after rebasing on top of develop.

**Pull request closed without merging**